# Credit Risk Probability of Default - Data Exploration

### Packages and Libraries 

In [1]:
# importing libraries and packages
import pandas as pd
import numpy as np


# visualization libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

import time as time 


In [2]:
pd.set_option('display.max_columns', None) # display max number of columns
pd.set_option('display.max_rows',None)

--------------------------------------------------------------------------------------------------------------------------------

### Loading the dataset

In [3]:
# Loading the dataset in the work environment
dataset_loan = pd.read_csv('./data/raw/Loan_Dataset.csv', low_memory = False)

# viewing a sample rows from the dataset
dataset_loan.sample(5)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
399792  NaN        NaN      27000        27000          27000.0   36 months   
184484  NaN        NaN      40000        40000          40000.0   60 months   
1712993 NaN        NaN      35000        35000          35000.0   36 months   
944239  NaN        NaN      20000        20000          19975.0   36 months   
1517157 NaN        NaN       8000         8000           8000.0   36 months   

         int_rate  installment grade sub_grade                 emp_title  \
399792      11.98       896.53     B        B5          Registered nurse   
184484      12.73       904.61     B        B5              Unit Manager   
1712993     14.30      1201.33     C        C1                  Director   
944239       9.17       637.58     B        B2  Manager of Annual Giving   
1517157     15.05       277.52     C        C4                       DSP   

        emp_length home_ownership  annual_inc verification_status   issue_d  \
399792     2 years           RENT     95000.0            Verified  Mar-2018   
184484   10+ years       MORTGAGE     90500.0            Verified  Aug-2018   
1712993    3 years       MORTGAGE    155000.0     Source Verified  Oct-2013   
944239     4 years       MORTGAGE     78000.0     Source Verified  Dec-2015   
1517157    6 years           RENT     32000.0     Source Verified  Aug-2017   

        loan_status pymnt_plan  url desc             purpose  \
399792      Current          n  NaN  NaN  debt_consolidation   
184484      Current          n  NaN  NaN         credit_card   
1712993  Fully Paid          n  NaN  NaN  debt_consolidation   
944239   Fully Paid          n  NaN  NaN         credit_card   
1517157     Current          n  NaN  NaN  debt_consolidation   

                           title zip_code addr_state    dti  delinq_2yrs  \
399792        Debt consolidation    112xx         NY  21.34          0.0   
184484   Credit card refinancing    245xx         VA  26.65          0.0   
1712993         Debt reconstruct    220xx         VA  20.79          1.0   
944239   Credit card refinancing    601xx         IL  30.38          0.0   
1517157       Debt consolidation    971xx         OR  23.11          0.0   

        earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
399792          Jul-2012             0.0                     NaN   
184484          Sep-2006             1.0                     NaN   
1712993         Feb-1983             0.0                    15.0   
944239          Jul-1999             1.0                     NaN   
1517157         Nov-2008             0.0                     NaN   

         mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
399792                      NaN      18.0      0.0      25392        89.7   
184484                      NaN      11.0      0.0      25801        70.9   
1712993                     NaN      11.0      0.0      33464        80.6   
944239                      NaN      12.0      0.0      39844        65.7   
1517157                     NaN       5.0      0.0       2641        20.3   

         total_acc initial_list_status  out_prncp  out_prncp_inv  \
399792        19.0                   f   19748.44       19748.44   
184484        29.0                   w   37040.83       37040.83   
1712993       31.0                   f       0.00           0.00   
944239        43.0                   w       0.00           0.00   
1517157       11.0                   w    4446.79        4446.79   

          total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
399792    9825.890000          9825.89          7251.56        2574.33   
184484    5399.370000          5399.37          2959.17        2440.20   
1712993  43247.573170         43247.57         35000.00        8247.57   
944239   22958.222009         22929.52         20000.00        2958.22   
1517157   4988.670000          4988.67          3553.21        1435.46   

         total_rec_late_fee  recov

Loan dataset has numerical and categorical features along with few columns with missing values. We will be going through these features one by one to identify their importance and select the most revelvant features for training the model. 

Let's know more about the loan dataset, before proceeding ahead.

In [12]:
# checking the shape of the dataset
print(f"Loan dataset has a shape of {dataset_loan.shape}")

Loan dataset has a shape of (2260668, 145)


So in total we have 145 features with +2 Million observation in the raw dataset which needs to be cleaned and modified according to our observations and requirement for model training and evaluation for a more generalized model which is able to customer default rate in advance before the issuing the loan from the client.  

In [13]:
# basic informaiton about the dataset
dataset_loan.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260668 entries, 0 to 2260667
Columns: 145 entries, id to settlement_term
dtypes: float64(105), int64(4), object(36)
memory usage: 5.8 GB


Because of the large number of feature dataframe.info() is not showing the detail breakdown of each feature and it's datatype. 

But we can conculde from the output that that we have 4 features with int64, 105 features with float64 and 36 features that are of object datatypes. 

We will have to check one by one these features for their datatype and convert or make changes accordingly. 

Before we will have to make a fucntion that will help to see the various other information about the no. of observations in the dataset, unique values in the dataset, percentage of missing values and so on, before proceeding ahead. 

**Task**
   - Duplicate values.
   - Correct Datatypes (**To be handle during EDA**)
   - Converting feature into more meaningful format. (**Handle During EDA**)
   - Missing values.

**Check for Duplicated values**

Duplicate rows can cause bias in model training for Porbability of Default(PD), impacting our model performance as they effectively give more/increase weight to certain samples.They may also indicate data collection issues.

Identifying and removing them from the dataset, helps to ensure data integrity.


In [5]:
# check for duplicated values in the loan dataset
dataset_loan.duplicated().sum()

0

As we can see we don't have any duplicate values in the loan dataset.

**Checking for Null Values/ Missing Values in the loan dataset**

Missing values can lead to error during data processing and may cause bias on inaccurate predictions in machine learning models. Detecting them early allows us to decide on appropriate handling methods, such as imputation or removal to maintin data quality.

Because of large number of feature in the loan dataset, we will be creating a function that will help to list all the features along with non-null values, null values count, percent of null values and datatype of these features as info() is not able to show all the 145 features. 

In [4]:
def dataset_meta_data(dataset):
    
    # dataframe to store the meta information about each features
    dataset_meta = pd.DataFrame(columns = ['Feature_name','Non_Null_Values',
                                           'Null_Values','%_Null_Values',
                                           'Unique_Values','Data_Type'])
    # length of the dataset
    l = len(dataset)
    
    # loop to collect meta data for each feature and append to dataset_meta
    for col in dataset.columns:
        
        # know the null values in the feature
        null_value_count = dataset[col].isnull().sum()
        
        # append the meta data about each feature
        dataset_meta.loc[len(dataset_meta),:] = [col,
                                                 dataset[col].notnull().sum(),
                                                 null_value_count,
                                                 round(null_value_count*100/l,2),
                                                 dataset[col].nunique(),
                                                 dataset[col].dtype
                                                ]
        
        
        
    # return the dataset_meta
    return dataset_meta    

In [8]:
dataset_meta_desc = dataset_meta_data(dataset_loan)

# print the meta data obtained from the dataset_meta_data() function
dataset_meta_desc

Feature_name Non_Null_Values Null_Values  \
0                                            id               0     2260668   
1                                     member_id               0     2260668   
2                                     loan_amnt         2260668           0   
3                                   funded_amnt         2260668           0   
4                               funded_amnt_inv         2260668           0   
5                                          term         2260668           0   
6                                      int_rate         2260668           0   
7                                   installment         2260668           0   
8                                         grade         2260668           0   
9                                     sub_grade         2260668           0   
10                                    emp_title         2093699      166969   
11                                   emp_length         2113761      146907   
12                               home_ownership         2260668           0   
13                                   annual_inc         2260664           4   
14                          verification_status         2260668           0   
15                                      issue_d         2260668           0   
16                                  loan_status         2260668           0   
17                                   pymnt_plan         2260668           0   
18                                          url               0     2260668   
19                                         desc          126067     2134601   
20                                      purpose         2260668           0   
21                                        title         2237343       23325   
22                                     zip_code         2260667           1   
23                                   addr_state         2260668           0   
24                                          dti         2258957        1711   
25                                  delinq_2yrs         2260639          29   
26                             earliest_cr_line         2260639          29   
27                               inq_last_6mths         2260638          30   
28                       mths_since_last_delinq         1102166     1158502   
29                       mths_since_last_record          359156     1901512   
30                                     open_acc         2260639          29   
31                                      pub_rec         2260639          29   
32                                    revol_bal         2260668           0   
33                                   revol_util         2258866        1802   
34                                    total_acc         2260639          29   
35                          initial_list_status         2260668           0   
36                                    out_prncp         2260668           0   
37                                out_prncp_inv         2260668           0   
38                                  total_pymnt         2260668           0   
39                              total_pymnt_inv         2260668           0   
40                              total_rec_prncp         2260668           0   
41                                total_rec_int         2260668           0   
42                           total_rec_late_fee         2260668           0   
43                                   recoveries         2260668           0   
44                      collection_recovery_fee         2260668           0   
45                                 last_pymnt_d         2258242        2426   
46                              last_pymnt_amnt         2260668           0   
47                                 next_pymnt_d          957061     1303607   
48                           last_credit_pull_d         2260595          73   
49                   collections_12_mths_ex_med         2260523         145   
50    

As we can see from the above table that there are too many columns with missing values and few of the categorical features with too many categories. We will first have to see deal with the null values in the features and set a threshold for the percent of null values that we can use for the features. 

For setting the threshold for percent of null values in the feature, let's first sort the dataset_meta_desc for better understanding about the null values.

We will view the columns with null values more than 1% as according to dataset_meta_desc we have few features with less than or close to 0%(i.e 100-2000 values missing) and won't make sense to view for this analysis.

In [23]:
# sorting the detaset_meta_desc better understanding of percent of null values 
dataset_meta_desc[dataset_meta_desc['%_Null_Values']>1].sort_values(by = ['%_Null_Values'], ascending = False)

Feature_name Non_Null_Values Null_Values  \
0                                            id               0     2260668   
18                                          url               0     2260668   
1                                     member_id               0     2260668   
134  orig_projected_additional_accrued_interest            8426     2252242   
131                             hardship_length           10613     2250055   
124                             hardship_reason           10613     2250055   
125                             hardship_status           10613     2250055   
126                               deferral_term           10613     2250055   
127                             hardship_amount           10613     2250055   
128                         hardship_start_date           10613     2250055   
129                           hardship_end_date           10613     2250055   
130                     payment_plan_start_date           10613     2250055   
133                        hardship_loan_status           10613     2250055   
132                                hardship_dpd           10613     2250055   
135              hardship_payoff_balance_amount           10613     2250055   
136                hardship_last_payment_amount           10613     2250055   
123                               hardship_type           10613     2250055   
139                   debt_settlement_flag_date           33056     2227612   
140                           settlement_status           33056     2227612   
141                             settlement_date           33056     2227612   
142                           settlement_amount           33056     2227612   
143                       settlement_percentage           33056     2227612   
144                             settlement_term           33056     2227612   
121         sec_app_mths_since_last_major_derog           35942     2224726   
116                          sec_app_revol_util          106184     2154484   
112                    sec_app_earliest_cr_line          108021     2152647   
111                             revol_bal_joint          108020     2152648   
113                      sec_app_inq_last_6mths          108021     2152647   
114                            sec_app_mort_acc          108021     2152647   
115                            sec_app_open_acc          108021     2152647   
117                         sec_app_open_act_il          108021     2152647   
118                       sec_app_num_rev_accts          108021     2152647   
119            sec_app_chargeoff_within_12_mths          108021     2152647   
120          sec_app_collections_12_mths_ex_med          108021     2152647   
55                    verification_status_joint          115730     2144938   
53                             annual_inc_joint          120710     2139958   
54                                    dti_joint          120706     2139962   
19                                         desc          126067     2134601   
29                       mths_since_last_record          359156     1901512   
86                     mths_since_recent_bc_dlq          519701     1740967   
50                  mths_since_last_major_derog          580775     1679893   
88               mths_since_recent_revol_delinq          740359     1520309   
47                                 next_pymnt_d          957061     1303607   
28                       mths_since_last_delinq         1102166     1158502   
65                                      il_util         1191818     1068850   
63                           mths_since_rcnt_il         1350744      909924   
69                                     all_util         1394320      866348   
68                                   max_bal_bc         1394539      866129   
67                                  open_rv_24m         1394539      866129   
71                                       inq_fi         1394539      866129   
72    

We can see that there is a clear distinction in the percent of null values in the loan dataset.

**Observation**

   - **100%** - These are either featues with either unique values or descriptive features like url which won't help in model training.
   
       **Features (`id`, `url`, `member_id`)**
    
   - **98.41-99.53** -> These features are related to depth settlement or hardship features which is out of scope for this model as of now. We can try to make a different model for predicting loan status during depth settlement and hardship type. We can seperate these values from the dataset.
   
       **Features - (`sec_app_mths_since_last_major_derog, debt_settlement_flag_date, settlement_status, settlement_date, settlement_amount, settlement_percentage, settlement_term`)**
       
       
    
   - **94 to 95.22 %** -> These features are mostly talking about the second applicant in the joint application for the loan. Most of the values present will be useful for defining the models for a joint application model. But currently we are making a model for both single and joint application.
   
       **Features - (`desc, annual_inc_joint, dti_joint, verification_status_joint, revol_bal_joint, sec_app_earliest_cr_line, sec_app_inq_last_6mths, sec_app_mort_acc, sec_app_open_acc, sec_app_revol_util, sec_app_open_act_il, sec_app_num_rev_accts, sec_app_chargeoff_within_12_mths, sec_app_collections_12_mths_ex_med`)**
       
    
   - **40.25 to 84.11** -> these features are various information about the customers which is linked to the current loan/ credit provided by the client. As we are making model for predicting the loan default before the issusing the loan we can drop these features.
   
       **Features - (`mths_since_last_delinq, mths_since_last_record, next_pymnt_d, mths_since_last_major_derog, mths_since_rcnt_il, il_util, mths_since_recent_bc_dlq, mths_since_recent_revol_delinq`)**
       
       
    
   - **38.1 %** -> these features talk about the final conditions of the borrowers and have to be evaluated. 
   
       **Features - (`open_acc_6m, open_act_il, open_il_12m, open_il_24m, total_bal_il, open_rv_12m, open_rv_24m, max_bal_bc, all_util, inq_fi, total_cu_tl, inq_last_12m`)**
   
    
   - **1% to 13.07%** -> these features are from various aspect and has to be evaluated one by one and selected according to the importance towards the task at hand.
   
       **Features - (`emp_title, emp_length, title, tot_coll_amt, tot_cur_bal, total_rev_hi_lim, acc_open_past_24mths, avg_cur_bal, bc_open_to_buy, bc_util, mo_sin_old_il_acct, mo_sin_old_rev_tl_op, mo_sin_rcnt_rev_tl_op, mo_sin_rcnt_tl, mort_acc, mths_since_recent_bc, mths_since_recent_inq, num_accts_ever_120_pd, num_actv_bc_tl, num_actv_rev_tl, num_bc_sats, num_bc_tl, num_il_tl, num_op_rev_tl, num_rev_accts, num_rev_tl_bal_gt_0, num_sats, num_tl_120dpd_2m, num_tl_30dpd, num_tl_90g_dpd_24m, num_tl_op_past_12m, pct_tl_nvr_dlq, percent_bc_gt_75, tot_hi_cred_lim, total_bal_ex_mort, total_bc_limit, total_il_high_credit_limit`)**

In [34]:
# knowing the number of columns according to null values segregation. 

# features with 100% null values
null_values_100 = dataset_meta_desc[dataset_meta_desc["%_Null_Values"]==100].iloc[:,0]
print(f'Features with 100% null values are {", ".join(null_values_100)} and count is {len(null_values_100.values)} .')

print("\n")
# features with 98.41 to 99.53% null values 
null_values_98_99 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>98) & (dataset_meta_desc["%_Null_Values"]<99)].iloc[:,0]
print(f'Features with 98.41 to 99.53% null values are {", ".join(null_values_98_99)} and count is {len(null_values_98_99.values)} .')

print("\n")
# features with 94 to 96% null values 
null_values_94_96 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>94) & (dataset_meta_desc["%_Null_Values"]<96)].iloc[:,0]
print(f'Features with 94 to 96% null values are {", ".join(null_values_94_96)} and count is {len(null_values_94_96.values)} .')

print("\n")
# features with 40 to 85% null values 
null_values_40_85 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]>40) & (dataset_meta_desc["%_Null_Values"]<85)].iloc[:,0]
print(f'Features with 40 to 85%  null values are {", ".join(null_values_40_85)} and count is {len(null_values_40_85.values)} .')

print("\n")
# features with 38.1% null values 
null_values_38 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]<39) & (dataset_meta_desc["%_Null_Values"]>30)].iloc[:,0]
print(f'Features with 38.1% null values are {", ".join(null_values_38)} and count is {len(null_values_38.values)} .')


print("\n")
# features less than 30% null values 
null_values_30 = dataset_meta_desc[(dataset_meta_desc["%_Null_Values"]<30) & (dataset_meta_desc["%_Null_Values"]>1)].iloc[:,0]
print(f'Features with 30% null values are {", ".join(null_values_30)} and count is {len(null_values_30.values)} .')



Features with 100% null values are id, member_id, url and count is 3 .


Features with 98.41 to 99.53% null values are sec_app_mths_since_last_major_derog, debt_settlement_flag_date, settlement_status, settlement_date, settlement_amount, settlement_percentage, settlement_term and count is 7 .


Features with 94 to 96% null values are desc, annual_inc_joint, dti_joint, verification_status_joint, revol_bal_joint, sec_app_earliest_cr_line, sec_app_inq_last_6mths, sec_app_mort_acc, sec_app_open_acc, sec_app_revol_util, sec_app_open_act_il, sec_app_num_rev_accts, sec_app_chargeoff_within_12_mths, sec_app_collections_12_mths_ex_med and count is 14 .


Features with 40 to 85%  null values are mths_since_last_delinq, mths_since_last_record, next_pymnt_d, mths_since_last_major_derog, mths_since_rcnt_il, il_util, mths_since_recent_bc_dlq, mths_since_recent_revol_delinq and count is 8 .


Features with 38.1% null values are open_acc_6m, open_act_il, open_il_12m, open_il_24m, total_bal_il, open_rv

**Observation**
    
    - Large number of features are having missing values as seen above.
    - Features could be divided according to the bins of percentage of missing values in that particular features.
    
  **[Note]** Most of these features are kpi that are tracked after the issusing of the loan or credit. So it's important to identify these columns and removing them from the dataset before we proceed with the EDA.
    
**Task**

   - Identify the Target Column
   - Identify the feature that are tracked after the issuance of the loan. **[Feature Selection]**
   - Remove these features which are tracked after issusing loan. [**Feature Selection**]
   - Drop the features with too many Null values. [**Feature Selection**]

--------------------------------------------------------------------------------------------------------------------------------

### Target Variable

We are trying to predict the loan_status variable in the load dataset.

Let's see the value distribution in loan_status variable.

In [8]:
# value distribution of variable loan_status

dataset_loan['loan_status'].value_counts()

Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: loan_status, dtype: int64

According to our business proble, we have to predict the whether the borrower will default before the issuing of the loan or credit line to the customers. So we are only predicting for two categories i.e **Fully Paid and Charged Off** . 

Therefore we will only keep observations with loan_status "Fully Paid" or "Charged Off".

In [9]:
# removing other categories in loan status other than Fully Paid or Charged Off
dataset_loan = dataset_loan.loc[dataset_loan['loan_status'].isin(['Fully Paid','Charged Off'])]



In [11]:
# checking the value distribution between "Fully Paid" and "Charged Off"
dataset_loan['loan_status'].value_counts(dropna = False)

Fully Paid     1041952
Charged Off     261655
Name: loan_status, dtype: int64

In [12]:
# checking the normalized distribution between Fully Paid and Charged Pff
dataset_loan['loan_status'].value_counts(normalize = True, dropna = True)

Fully Paid     0.799284
Charged Off    0.200716
Name: loan_status, dtype: float64

**Observations**
 - Close to 80% of the loans are Fully Paid and around 20% of the loans are Charged Off.
 - We will have to handle the unbalance dataset that we have before training either using cross validation or SMOTE method for optimal performance for our model.
 
**Task**
  - Handle imbalance between the two categories using 
      - Cross Validation method. 
      - SMOTE method.

-------------------------------------------------------------------------------

We have done the initial checks for our loan dataset and according to the observations we have few of the tasks that needs to be performed as we move forward. 

We will move with Feature Selection and EDA in the below section

### Feature Selection

### Saving intermediate data

In [ ]:
dataset_loan.to_csv('./data/proceesed/temp_processed_data_1.csv')